working/1426-extension-types/feature-specification.md

```diff
@@ -0,0 +1,917 @@
+  ListSize<String> xs = <String>['Hello']; // OK, upcast.
+  print(xs); // OK, `toString()` available on Object?.
+  print("Size: ${xs.size}. Front: ${xs.front()}"); // OK.
+  xs[0]; // Error, `operator []` not a member of `ListSize`.
+
+  List<ListSize<String>> ys = [xs]; // OK.
+  List<List<String>> ys2 = ys; // Error, downcast.
+  ListSize<ListSize<Object>> ys3 = ys; // OK.
+  ys[0].front(); // OK.
+  ys3.front().front(); // OK.
+  ys as List<List<String>>; // `ys` is promoted, succeeds at run time.
+}
+```
+
+
+## Syntax
+
+The rule for `<extensionDeclaration>` in the grammar is replaced by the
+following:
+
+```ebnf
+<extensionDeclaration> ::=
+  'protected'? 'extension' ('type'? <typeIdentifier> <typeParameters>?)?
+      <extensionShowHidePart> 'on' <type> <interfaces>? '{'
+    (<metadata> <extensionMemberDefinition>)*
+  '}'
+
+<extensionShowHidePart> ::=
+  <extensionShowClause>? <extensionHideClause>?
+
+<extensionShowClause> ::= 'show' <extensionShowHideList>
+
+<extensionHideClause> ::= 'hide' <extensionShowHideList>
+
+<extensionShowHideList> ::=
+  <extensionShowHideElement> (',' <extensionShowHideElement>)*
+
+<extensionShowHideElement> ::=
+  <type> | <identifier> | 'operator' <operator>
+```
+
+*In the rule `<extensionShowHideElement>`, note that `<type>` derives
+`<typeIdentifier>`, which makes `<identifier>` nearly redundant. However,
+`<identifier>` is still needed because it includes some strings that cannot
+be the name of a type but can be the basename of a member, e.g., the
+built-in identifiers.*
+
+
+## Static Analysis
+
+Assume that _E_ is an extension declaration of the following form:
+
+```dart
+extension Ext<X1 extends B1, .. Xk extends Bk> on T {
+  ... // Members
+}
+```
+
+*Note that `extension type` declarations and other enhanced forms are
+discussed in later sections. The properties specified here are valid for
+those forms as well, except for the differences which are specified
+explicitly.*
+
+It is then allowed to use `Ext<S1, .. Sk>` as a type: It can occur as the
+declared type of a variable or parameter, as the return type of a function
+or getter, as a type argument in a type, or as the on-type of an extension.
+
+*In particular, it is allowed to create a new instance where one or more
+extension types occur as type arguments.*
+
+When `k` is zero, `Ext<S1, .. Sk>` simply stands for `Ext`, a non-generic
+extension. When `k` is greater than zero, a raw occurrence `Ext` is treated
+like a raw type: Instantiation to bound is used to obtain the omitted type
+arguments.
+
+We say that the static type of said variable, parameter, etc. _is the
+extension type_ `Ext<S1, .. Sk>`, and that its static type _is an extension
+type_.
+
+If `e` is an expression whose static type is the extension type
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>,
+then a member access like `e.m(args)` is treated as
+<code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(v).m(args)</code>
+where `v` is a fresh variable whose declared type
```

Comment:
... and `v` is bound to the value of `e`, which is evaluated at ... some point ...?

I'd just say that it's equivalent to <code>Ext<S<sub>1</sub>, .. S<sub>k</sub>>(e).m(args)</code> where the static type of `e` is the `on` type of `Ext<S<sub>1</sub>, .. S<sub>k</sub>></code>.

Reply:
This section is about the static analysis, and the part about evaluating `e` to an object `o` and binding `v` to `o` is in the section about the dynamic semantics.

We could say "where the static type of `e` is ...", but I don't think we have any other situation where we force a static type onto an existing expression by fiat. I think it could be a source of confusion ("so what if `e` doesn't actually have that type, is it an error, or do we add an implicit cast, ...?").

In an earlier version of this document, I did not have `v`, and I used `e as T` instead (where `T` is the on-type). This will actually produce the intended static analysis, because the argument to `E(...)` has the required static type.

However, that gave rise to a another issue: It could make the reader think that the cast could fail. (In fact, it is guaranteed to succeed, and could hence be a no-op at run time).

I've changed the text to use an `as` expression again, and added commentary to indicate that this cast won't fail and does not take time (because it is statically known to succeed).

Reply:
(This part was extensively rewritten; resolving this thread)